### PR TITLE
feat: Initial implementation of `cp-connector` library

### PR DESCRIPTION
**Pull request by warber**
THis PR contains the first implementation of `cp-connector` a library that can be used instead of the distributor to implement keptn services. The first POC of using this lib with the lighthouse service can be found here #7334

Note that `internal.go`, `log.go` as well as `eventmatcher.go` was copied from existing modules, hence the code duplication warning from sonar. Eventually, this code can be moved into a "common" module, as it planned with https://github.com/keptn/keptn/issues/7079

### CODE REVIEW

**Review by RealAnna — COMMENTED** [review]
now LGTM but I think @bacherfl should also check it 😄

**Comment by mowies** [comment]
Maybe it would be good to already introduce a new scope for our semantic PR checking in this PR so that further features for the library are grouped in the changelog correctly.

File: lib-cp-connector/pkg/api/internal.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+var DefaultInClusterAPIMappings = InClusterAPIMappings{
+	ConfigurationService: "configuration-service:8080",
+	ShipyardController:   "shipyard-controller:8080",
+	ApiService:           "api-service:8080",
+	SecretService:        "secret-service:8080",
+	MongoDBDatastore:     "mongodb-datastore:8080",
+}
+
+// NewInternal creates a new InternalAPISet usable for calling keptn services from within the control plane
+func NewInternal(client *http.Client, apiMappings ...InClusterAPIMappings) (*InternalAPISet, error) {
```
Comment (bacherfl) [inline lib-cp-connector/pkg/api/internal.go:56]: given that the package only contains this component, should we rename this to just `New()`? Or will this eventually also include an `ExternalAPISet`?

File: lib-cp-connector/pkg/controlplane/controlplane.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	for _, subscription := range cp.currentSubscriptions {
+		if subscription.Event == eventUpdate.MetaData.Subject {
+			matcher := NewEventMatcherFromSubscription(subscription)
+			if matcher.Matches(eventUpdate.KeptnEvent) {
+				if err := integration.OnEvent(context.WithValue(ctx, EventSenderKey, cp.eventSource.Sender()), eventUpdate.KeptnEvent); err != nil {
+					if errors.Is(err, ErrEventHandleFatal) {
+						cp.logger.Errorf("Fatal error during handling of event: %v", err)
+						return err
+					}
+					if errors.Is(err, ErrEventHandleIgnore) {
```
Comment (bacherfl) [inline lib-cp-connector/pkg/controlplane/controlplane.go:80]: should we add a default case here as well, to also log errors other than `EventHandleFatal` and `ErrEventHandleIgnore`?

File: lib-cp-connector/pkg/api/internal.go (hunk 1 above)
Comment (warber) [reply lib-cp-connector/pkg/api/internal.go:56]: It *could* be that there will be a `NewExternal` function besides this one which will just use the regular `APISet` provided by `go-utils` https://github.com/keptn/go-utils/blob/24667368a1594c17cfa725b88a97d19ddfef109e/pkg/api/utils/client.go#L146

Comment (warber) [reply lib-cp-connector/pkg/api/internal.go:56]: Update: agreed with @bacherfl  to keep `NewInternal` for now

File: lib-cp-connector/pkg/controlplane/controlplane.go (hunk 2 above)
Comment (warber) [reply lib-cp-connector/pkg/controlplane/controlplane.go:80]: also the question is, what should be the default behavior?
should the user actively need to decide that it is a fatal error or should she actively decide that it is an error that can be ignored?
i.e., **any** error will be treated to be ignoreable, while `ErrEventHandleFatal` will be treated as fatal
or
**any** error will be treated to be fatal, while `ErrEventHandleIgnore` will be treated as ignore

**Comment by bacherfl** [comment]
also LGTM from my side, nice work!

**Comment by mowies** [comment]
Please add the new scope for the library here: https://github.com/keptn/keptn/blob/master/.github/workflows/validate-semantic-pr.yml

**Review by thisthat — APPROVED** [review]

**Review by thisthat — APPROVED** [review]

**Pull request merged**